internal/grpc/commander/connection.go

```diff
@@ -142,9 +398,111 @@ func (c *connection) register(nginxID, systemID string) error {
+// Figure out how to remove this without causing errors in the agent.
+func (c *connection) receiveFromUploadServer(server proto.Commander_UploadServer) error {
+	c.logger.Info("Upload request")
+
+	for {
+		// Recv blocks until it receives a message into or the stream is
+		// done. It returns io.EOF when the client has performed a CloseSend. On
+		// any non-EOF error, the stream is aborted and the error contains the
+		// RPC status.
+		_, err := server.Recv()
+
+		if err != nil && !errors.Is(err, io.EOF) {
+			c.logger.Error(err, "upload receive error")
+			return err
+		}
+
+		c.logger.Info("Received chunk from upload channel")
+
+		if errors.Is(err, io.EOF) {
+			c.logger.Info("Upload completed")
+			return server.SendAndClose(&proto.UploadStatus{Status: proto.UploadStatus_OK})
+		}
+	}
+}
+
+func (c *connection) handleDataplaneStatus(ctx context.Context, status *proto.DataplaneStatus) {
+	// Right now, we only care about AgentActivityStatuses that contain NginxConfigStatuses.
+	if status.GetAgentActivityStatus() != nil {
+		for _, activityStatus := range status.GetAgentActivityStatus() {
+			if cfgStatus := activityStatus.GetNginxConfigStatus(); cfgStatus != nil {
+				c.handleNginxConfigStatus(ctx, cfgStatus)
+			}
+		}
+	}
+}
+
+func (c *connection) handleNginxConfigStatus(ctx context.Context, status *proto.NginxConfigStatus) {
+	c.logger.Info("Received nginx config status", "status", status.Status, "message", status.Message)
+	// If status is pending then we need to wait for the next status update
+	if status.Status == proto.NginxConfigStatus_PENDING {
+		return
+	}
+
+	applyStatus := configApplyStatusSuccess
+	if status.Status == proto.NginxConfigStatus_ERROR {
+		applyStatus = configApplyStatusFailure
+	}
+
+	res := configApplyResponse{
+		correlationID: status.CorrelationId,
+		status:        applyStatus,
+		message:       status.Message,
+	}
+
+	c.sendConfigApplyResponse(ctx, res)
+}
+
+func (c *connection) handleNginxConfigResponse(ctx context.Context, res *proto.NginxConfigResponse) {
+	status := res.Status
+
+	c.logger.Info("Received nginx config response", "status", status.Status, "message", status.Message)
+
+	// We only care about ERROR status because it indicates that the config apply action is complete.
+	// An OK status can indicate that the config apply action is still in progress or that it is complete. However,
+	// the Agent will send a DataplaneStatus update on a successful config apply, so we don't need to handle it here.
+	// We handle the error case here, because in some cases, the Agent will not send a DataplaneStatus update on a
+	// failed config apply.
+	if status.Status != proto.CommandStatusResponse_CMD_ERROR {
+		return
+	}
+
+	car := configApplyResponse{
+		status:  configApplyStatusFailure,
+		message: status.Error,
+	}
+
+	c.sendConfigApplyResponse(ctx, car)
+}
+
+func (c *connection) sendConfigApplyResponse(ctx context.Context, response configApplyResponse) {
```

Comment:
"send" is a bit misleading here, because we're not sending anything to the agent